### PR TITLE
rustdoc: Remove distinction between "regular" and "collapsed" docs

### PR DESCRIPTION
**Pull request by camelid**
There actually seems to be no difference.

r? @GuillaumeGomez 
cc @jyn514

### CODE REVIEW

**Comment by camelid** [comment]
Might as well see if the Crate shrinkage improved perf.

@bors try @rust-timer queue

File: src/librustdoc/clean/types.rs
Comment (jyn514) [inline src/librustdoc/clean/types.rs]: This is not the same behavior. The `FromIterator` impl below is missing the `new_frag.kind != ori.kind || new_frag.parent_module != ori.parent_module` check.

I'm not sure if it matters in practice, but it seems like something we should have tests for ... could you try and find one or two, or add a new one if not?

File: src/librustdoc/clean/types.rs
Comment (jyn514) [inline src/librustdoc/clean/types.rs:1082]: not necessarily in this pr, but it would be nice to remove the `Option` here

**Review by jyn514 — COMMENTED** [review]
I'm not sure this has the same behavior ... some things are obviously a win like turning `FromIterator` into a function, but I would like to see more examples of the behavior for different doc types before and after.

File: src/librustdoc/clean/types.rs
Comment (GuillaumeGomez) [inline src/librustdoc/clean/types.rs:127]: Nice!

File: src/librustdoc/clean/types.rs
Comment (GuillaumeGomez) [reply src/librustdoc/clean/types.rs:1082]: I think we check the `is_some` to generate differently the doc block. From my point of view, this forces us to keep this check so might be better to keep it.

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
Let's restart the perf check:

@bors try @rust-timer queue

File: src/librustdoc/clean/types.rs
Comment (camelid) [reply src/librustdoc/clean/types.rs]: Before merging collapsed and un-collapsed, I tried removing one part of this conditional and all the tests passed. I'll try removing the whole conditional too (with no other changes) to see what happens.

I agree it'd be good to have more test coverage. I'll see if I can come up with some test cases.

**Comment by camelid** [comment]
Ugh, it'll be nice to have `cargo check --doc` 😅 

```
error: unresolved link to `DocFragments`
32468
   --> src/librustdoc/clean/types.rs:947:32
32469
    |
32470
947 | /// Collapse a collection of [`DocFragments`] into one string,
32471
    |                                ^^^^^^^^^^^^ no item named `DocFragments` in scope
32472
    |
32473
    = note: `-D rustdoc::broken-intra-doc-links` implied by `-D warnings`
32474
    = help: to escape `[` and `]` characters, add '\' before them like `\[` or `\]`
32475
```

File: src/librustdoc/clean/types.rs
Comment (camelid) [reply src/librustdoc/clean/types.rs]: All tests pass with master + this diff:

```diff
diff --git a/src/librustdoc/clean/types.rs b/src/librustdoc/clean/types.rs
index 41ebf270ba6..5a02a6c6259 100644
--- a/src/librustdoc/clean/types.rs
+++ b/src/librustdoc/clean/types.rs
@@ -1083,9 +1083,6 @@ fn update_need_backline(doc_strings: &mut Vec<DocFragment>) {
         let mut out = String::new();
         add_doc_fragment(&mut out, ori);
         for new_frag in iter {
-            if new_frag.kind != ori.kind || new_frag.parent_module != ori.parent_module {
-                break;
-            }
             add_doc_fragment(&mut out, new_frag);
         }
         if out.is_empty() { None } else { Some(out) }
```

So either that conditional is useless or we don't have any test coverage for it.

File: src/librustdoc/clean/types.rs
Comment (camelid) [reply src/librustdoc/clean/types.rs:1082]: I don't really understand why `None` is treated differently from `Some("")` though. What if someone writes `#[doc = ""]`? Won't that return `Some("")` instead of `None`? And why do we need to special-case the absence of doc attrs so much?

**Comment by GuillaumeGomez** [comment]
Very small improvements but overall we're moving in the right direction! :D

File: src/librustdoc/clean/types.rs
Comment (camelid) [reply src/librustdoc/clean/types.rs]: Adding enough test coverage around this code to know for sure that nothing is changing will take a long time. What about if I open a PR to remove this conditional, we land it, and wait to see if any regressions pop up? If there are any regressions, then we can do a quick revert since it's just a few lines. If we don't find any regressions, then we can move forward with a bigger change.

Comment (jyn514) [reply src/librustdoc/clean/types.rs]: Hmm, I'm ok with that, but let's wait until after the beta branch on friday.

**Comment by camelid** [comment]
Blocked on #91305.

File: src/librustdoc/clean/types.rs
Comment (camelid) [reply src/librustdoc/clean/types.rs]: Opened #91305.

File: src/librustdoc/clean/types.rs
Comment (jyn514) [reply src/librustdoc/clean/types.rs:1082]: Another difference between `doc_value` and `collapsed_doc_value` - `doc_value` returns None if the *combined* string is empty, `collapsed_doc_value` returns None if there are no *lines*. The difference matters if there's empty lines, like below:
```
///
///
///
pub fn foo() {}
```

Comment (jyn514) [reply src/librustdoc/clean/types.rs:1082]: @CraftSpider I see the following comment in `rustdoc-json-types`:
```
    /// The full markdown docstring of this item. Absent if there is no documentation at all,
    /// Some("") if there is some documentation but it is empty (EG `#[doc = ""]`).
    pub docs: Option<String>,
```
Is that an important distinction? Or can I just make that field a simple `String`?

**Comment by camelid** [comment]
The PR this has been blocked on, #91305, is on stable as of a week and a half ago. As of yet, I don't know of any regressions that have been reported. So, I think in a month or two, we should be able to merge this PR as well!

**Comment by Dylan-DPC** [comment]
@camelid this is waiting on a rebase

**Comment by camelid** [comment]
Thanks, it's on my list. I might not get to it for a bit since there are so many conflicts.

**Comment by JohnCSimon** [comment]
@camelid 
Ping from triage: I'm closing this due to inactivity, Please reopen when you are ready to continue with this. 
Note: if you do please open the PR **BEFORE** you push to it, else you won't be able to reopen - this is a quirk of github. 
Thanks for your contribution.

@rustbot label: +S-inactive

**Pull request closed without merging**
